File: problems/0008_string-to-integer-atoi/one-loop/solution.c
int myAtoi(char * str) {
    if (str[0] == '\0')
        return 0;
    size_t i = 0;
    for (; str[i] == ' '; ++i);
    long sign = 1;
    if (str[i] == '-') {
        ++i;
        sign = -1;
    } else if (str[i] == '+')
        ++i;
    long result = 0;
    for (char c = str[i]; c != '\0' && '0' <= c && c <= '9'; c = str[++i]) {
        result = result * 10 + sign * (c - '0');
        if (result < INT_MIN)
            return INT_MIN;
        else if (result > INT_MAX)
            return INT_MAX;
    }
    return result;
}

File: problems/0008_string-to-integer-atoi/one-loop/Solution.py
#!/usr/bin/python3

class Solution:
    def myAtoi(self, str: str) -> int:
        sLen = len(str)
        if sLen == 0:
            return 0
        i = 0
        while i < sLen and str[i] == ' ':
            i += 1
        sign = 1
        if i < sLen:
            if str[i] == '-':
                i += 1
                sign = -1
            elif str[i] == '+':
                i += 1
        result = 0
        while i < sLen and '0' <= str[i] and str[i] <= '9':
            result = result * 10 + sign * (ord(str[i]) - ord('0'))
            if result < Solution.INT_MIN:
                return Solution.INT_MIN
            elif result > Solution.INT_MAX:
                return Solution.INT_MAX
            i += 1
        return result
    INT_MIN = -2**31
    INT_MAX = 2**31 - 1
